src/Microsoft.OData.Core/Json/JsonReader.cs

```diff
@@ -1261,19 +1544,682 @@ private bool ReadInput()
+
+        /// <summary>
+        /// Asynchronously parses a "value", that is an array, object or primitive value.
+        /// </summary>
+        /// <returns>A task that represents the asynchronous operation.
+        /// The value of the TResult parameter contains the node type to report to the user.</returns>
+        private async Task<JsonNodeType> ParseValueAsync()
+        {
+            Debug.Assert(
+                this.tokenStartIndex < this.storedCharacterCount && !IsWhitespaceCharacter(this.characterBuffer[this.tokenStartIndex]),
+                "The SkipWhitespaces wasn't called or it didn't correctly skip all whitespace characters from the input.");
+            Debug.Assert(this.scopes.Count >= 1 && this.scopes.Peek().Type != ScopeType.Object, "Value can only occur at the root, in array or as a property value.");
+
+            // Increase the count of values under the current scope.
+            this.scopes.Peek().ValueCount++;
+
+            char currentCharacter = this.characterBuffer[this.tokenStartIndex];
+            switch (currentCharacter)
+            {
+                case '{':
+                    // Start of object
+                    this.PushScope(ScopeType.Object);
+                    this.tokenStartIndex++;
+                    return JsonNodeType.StartObject;
+
+                case '[':
+                    // Start of array
+                    this.PushScope(ScopeType.Array);
+                    this.tokenStartIndex++;
+                    await this.SkipWhitespacesAsync().ConfigureAwait(false);
+                    this.canStream =
+                        this.characterBuffer[this.tokenStartIndex] == '"' ||
+                        this.characterBuffer[this.tokenStartIndex] == '\'' ||
+                        this.characterBuffer[this.tokenStartIndex] == 'n';
+                    return JsonNodeType.StartArray;
+
+                case '"':
+                case '\'':
+                    // String primitive value
+                    // Don't parse yet, as it may be a stream. Defer parsing until .Value is called.
+                    this.canStream = true;
+                    break;
+
+                case 'n':
+                    // Null value
+                    // Don't parse yet, as user may be streaming a stream. Defer parsing until .Value is called.
+                    this.canStream = true;
+                    break;
+
+                case 't':
+                case 'f':
+                    this.nodeValue = await this.ParseBooleanPrimitiveValueAsync()
+                        .ConfigureAwait(false);
+                    break;
+
+                default:
+                    // COMPAT 47: JSON number can start with dot.
+                    // The JSON spec doesn't allow numbers to start with ., but WCF DS does. We will follow the WCF DS behavior for compatibility.
+                    if (Char.IsDigit(currentCharacter) || (currentCharacter == '-') || (currentCharacter == '.'))

```

Comment:
Super minor potential performance change, let `'-'` and `'.'` be cases in the `switch` statement